content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,577 @@
+---
+title: Higher-order Functions in Javascript
+date: "2019-12-17T20:00:00.284Z"
+description: Learning and using higher-order functions in Javascript
+---
+
+# Higher-order functions in Javascript
+
+Higher-order functions, you've probably heard the term before. At first glance, it might sound scary, but don't worry.
+If you've been coding javascript for a while, you've probably used them many times!
+
+## Functional programming
+
+Before we start examining what a higher-order function is, let's take a step back and look at where they come from - functionals programming.
+Functional programming is one of many other programming paradigms (you might have heard of Object-Oriented Programming). Functional programming is a way of programming where you tell the computer **what** your program does, instead of
+**how**. It also emphasizes immutability, and treats functions as "first-class citizens".
+
+So what is a "first-class citizen" in programming, you might ask yourself?
+Well, turns out that first-class citizens are a special group of entities that enjoys some particular right and privileges:
+
+1. They can be stored in variables
+2. They can be passed as an argument to other procedures
+3. They can be returned as values from other procedures
+4. They can be incorporated into data structures
+
+Take a good look at numbers 2 and 3, and remember those two.
+
+This programming paradigm has many other characteristics, but that is a topic for another blog post.
+
+Javascript has some characteristics of a functional programming language, one of those being that it treats functions as guess what?
+That's right, first-class citizens!
+
+## Higher-order functions
+
+Remember numbers 2 and 3 from the list above? Well, those two describe a higher-order function (in our case, a procedure is another function).
+Putting it in words, a higher-order function is a function that does one or both of these:
+
+- It can take other functions as arguments
+- It can return other functions as arguments.
+
+Having support for higher-order functions opens a world of possibilities, such as currying, partial application, function composition. These are topics for other blog posts though.
+
+### Returning functions
+By now, you have read the word "function" about 20 times, but haven't seen a single one. Where is the action you might ask? Well, ask and you shall receive.
+
+
+```js
+const add = x => y => x + y
```

Comment:
I would prefer the `function` variant to be mentioned first and then write something like: `you can write this with a shorter notation using arrow functions...`